src/scaler/spreadscaler.rs

```diff
@@ -0,0 +1,1197 @@
+    pub config: ActorSpreadConfig,
+    store: S,
+}
+
+#[async_trait]
+impl<S: ReadStore + Send + Sync> Scaler for ActorSpreadScaler<S> {
+    type Config = ActorSpreadConfig;
+
+    async fn update_config(&mut self, config: Self::Config) -> Result<Vec<Command>> {
+        self.config = config;
+        self.reconcile().await
+    }
+
+    async fn handle_event(&self, event: &Event) -> Result<Vec<Command>> {
+        // NOTE(brooksmtownsend): We could be more efficient here and instead of running
+        // the entire reconcile, smart compute exactly what needs to change, but it just
+        // requires more code branches and would be fine as a future improvement
+        match event {
+            Event::ActorStarted(actor_started) => {
+                if actor_started.image_ref == self.config.actor_reference {
+                    self.reconcile().await
+                } else {
+                    Ok(Vec::new())
+                }
+            }
+            Event::ActorStopped(actor_stopped) => {
+                let actor_id = self.actor_id().await?;
+                if actor_stopped.public_key == actor_id {
+                    self.reconcile().await
+                } else {
+                    Ok(Vec::new())
+                }
+            }
+            Event::HostStopped(HostStopped { labels, .. })
+            | Event::HostStarted(HostStarted { labels, .. }) => {
+                // If the host labels match any spread requirement, perform reconcile
+                if self.config.spread_config.spread.iter().any(|spread| {
+                    spread.requirements.iter().all(|(key, value)| {
+                        labels.get(key).map(|val| val == value).unwrap_or(false)
+                    })
+                }) {
+                    Ok(Vec::new())
+                } else {
+                    self.reconcile().await
+                }
+            }
+            // No other event impacts the job of this scaler so we can ignore it
+            _ => Ok(Vec::new()),
+        }
+    }
+
+    async fn reconcile(&self) -> Result<Vec<Command>> {
+        let spread_requirements = compute_spread(&self.config.spread_config)?;
+        let hosts = self.store.list::<Host>(&self.config.lattice_id).await?;
+        let actor_id = self.actor_id().await?;
+        let actor = self
+            .store
+            .get::<Actor>(&self.config.lattice_id, &actor_id)
+            .await?;
+
+        // NOTE(brooksmtownsend) it's easier to assign one host per list of requirements than
+        // balance within those requirements. Users should be specific with their requirements
+        // as wadm is not responsible for ambiguity, a future scaler like a DaemonScaler could handle this
+
+        let commands = spread_requirements
+            .iter()
+            .filter_map(|(spread, count)| {
+                let eligible_hosts = eligible_hosts(&hosts, spread);
+                if let Some(first_host) = eligible_hosts.get(0) {
```

Comment:
Theoretically `eligible_hosts` _should_ return a different ordering of hosts every time because it is iterating over a hashmap to generate them. However, I worry about this favoring 1 host over another too much. For example, if I had two hosts that matched the requirement `foo: bar` and had those requirements on several manifests, I could get the same host multiple times for different models even though 2 match.

Now that I am thinking about this more, it could actually be a problem because the host you select is non-deterministic. On one call to reconcile, I could get Host A, and on a second call, I could get host B. We need to come up with a way to deterministically choose from a pool of hosts or spread across a pool of hosts. We _might_ be able to hold out on this until wadm 0.5, but we should talk about it. At a bare minimum, we should probably sort the vec of eligible hosts

Reply:
That is true, I thought one of my test cases covered this but the vec of spreads is ordered while the map of hosts isn't 🤔 

I liked the idea of it being random just to not favor one host too much, and I actually think non-determinism is okay here. Essentially here's what would possibly happen:

1. Define spread that asks for 12 replicas of an actor, 10 are running across the lattice on eligible hosts
2. We ask to start 2 more instances on host A
3. we start 2 more instances on host A
4. Next call to reconcile finds 12 actors on eligible hosts, so no commands needed
5. Sometime later, user deletes 5 instances from some host
6. We find 7 running actors on eligible hosts
7. We ask to start 5 more instances on host A, or maybe B, or maybe C
8. we start 5 more
9. next call to reconcile finds nothing to do

e.g. I think it's not a problem here and the randomness is kinda okay. Do you think it would cause problems?

Reply:
What if the count for this spread is across several nodes? Right now this only gets the count on the first node. Example logic

eligible hosts: [HostA, HostB]
replicas: 10

1. First reconcile chooses HostA to put 10 actors on it
2. 3 actors are stopped on HostA
3. Second reconcile schedules on HostB, sees a current count of actors of 0, schedules 10 actors on HostB
4. Now 20 replicas are running

Am I reading the logic right?

Reply:
After talking to @brooksmtownsend, I was reading the logic wrong. This block just selects a host and _then_ checks all running instances of the actor